sql/pgwire: TestParseClientProvidedSessionParameters failed

sql/pgwire.TestParseClientProvidedSessionParameters [failed](https://tanzanite.cluster.engflow.com/invocations/default/c3af9552-3cdb-4ea6-b57c-71957ce8d83e?testReportRun=22&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvc3FsL3Bnd2lyZTpwZ3dpcmVfdGVzdA==) on master @ [3c41c509a87cba7a1fd3f5cfdb0f6badb78e3704](https://github.com/cockroachdb/cockroach/commits/3c41c509a87cba7a1fd3f5cfdb0f6badb78e3704):

Fatal error:

```
panic: test timed out after 4m57s
running tests:
	TestParseClientProvidedSessionParameters (4m57s)
	TestParseClientProvidedSessionParameters/user_is_set_from_query (4m57s)

```
Stack: 

```
goroutine 24 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestParseClientProvidedSessionParameters
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestParseClientProvidedSessionParameters2373417980
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestParseClientProvidedSessionParameters/user_is_set_from_query
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=22</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestParseClientProvidedSessionParameters.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35879

--- a/pkg/sql/pgwire/BUILD.bazel
+++ b/pkg/sql/pgwire/BUILD.bazel
@@ -138,6 +138,7 @@ go_test(
         "//pkg/sql/pgwire/pgcode",
         "//pkg/sql/pgwire/pgerror",
         "//pkg/sql/pgwire/pgwirebase",
+        "//pkg/sql/pgwire/pgwirecancel",
         "//pkg/sql/randgen",
         "//pkg/sql/sem/eval",
         "//pkg/sql/sem/tree",

--- a/pkg/sql/pgwire/conn_test.go
+++ b/pkg/sql/pgwire/conn_test.go
@@ -1626,10 +1626,12 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 		t.Run(tc.desc, func(t *testing.T) {
 
 			var netConn net.Conn
-			wg := sync.WaitGroup{}
-			wg.Add(1)
+			clientDone := sync.WaitGroup{}
+			clientDone.Add(1)
+			serverReceivedConn := sync.WaitGroup{}
+			serverReceivedConn.Add(1)
 			go func(query string) {
-				defer wg.Done()
+				defer clientDone.Done()
 				ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
 				defer cancel()
 				url := fmt.Sprintf("%s&%s", baseURL, query)
@@ -1641,6 +1643,7 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 				// ignore the error because there is no answer from the server, we are
 				// interested in parsing session arguments only. We close the connection
 				// immediately, since getSessionArgs is blocking.
+				serverReceivedConn.Wait()
 				_ = netConn.Close()
 				_ = c.Close(ctx)
 
@@ -1652,10 +1655,15 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 					t.Error("pgconn asyncClose did not clean up on time")
 				}
 			}(tc.query)
-			// Wait for the client to connect and perform the handshake.
+			// Wait for the client to connect and perform the handshake. Use a
+			// closure so that the WaitGroup is marked as done even if there's a
+			// panic.
 			var args sql.SessionArgs
-			netConn, args, err = getSessionArgs(ln, true /* trustRemoteAddr */)
-			wg.Wait()
+			func() {
+				defer serverReceivedConn.Done()
+				netConn, args, err = getSessionArgs(ln, true /* trustRemoteAddr */)
+			}()
+			clientDone.Wait()
 			tc.assert(t, args, err)
 		})
 	}

--- a/pkg/sql/pgwire/conn_test.go
+++ b/pkg/sql/pgwire/conn_test.go
@@ -38,6 +38,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirecancel"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
@@ -561,15 +562,16 @@ func waitForClientConn(ln net.Listener) (net.Conn, error) {
 // together with session arguments or an error.
 func getSessionArgs(
 	ln net.Listener, trustRemoteAddr bool,
-) (conn net.Conn, _ sql.SessionArgs, err error) {
+) (netConn net.Conn, _ sql.SessionArgs, retErr error) {
 	for {
-		conn, err = ln.Accept()
+		var err error
+		netConn, err = ln.Accept()
 		if err != nil {
 			return nil, sql.SessionArgs{}, err
 		}
 
 		buf := pgwirebase.MakeReadBuffer()
-		_, err = buf.ReadUntypedMsg(conn)
+		_, err = buf.ReadUntypedMsg(netConn)
 		if err != nil {
 			return nil, sql.SessionArgs{}, err
 		}
@@ -585,15 +587,32 @@ func getSessionArgs(
 		if version != version30 {
 			return nil, sql.SessionArgs{}, errors.Errorf("unexpected protocol version: %d", version)
 		}
-
+		defer func() {
+			// Implement a fake pgwire connection handshake. Send the response
+			// after parsing the client-sent parameters.
+			c := &conn{conn: netConn}
+			c.msgBuilder.init(metric.NewCounter(metric.Metadata{}))
+			if err := c.authOKMessage(); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+				return
+			}
+			if err := c.bufferInitialReadyForQuery(pgwirecancel.BackendKeyData(0)); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+				return
+			}
+			if err := c.Flush(0); err != nil {
+				retErr = errors.CombineErrors(retErr, err)
+				return
+			}
+		}()
 		ctx := context.Background()
-		cp, err := parseClientProvidedSessionParameters(ctx, &buf, conn.RemoteAddr(), trustRemoteAddr,
+		cp, err := parseClientProvidedSessionParameters(ctx, &buf, netConn.RemoteAddr(), trustRemoteAddr,
 			false /* acceptTenantName */, false /* acceptSystemIdentityOption */)
 		if err != nil {
-			return conn, sql.SessionArgs{}, err
+			return netConn, sql.SessionArgs{}, err
 		}
 		args, err := finalizeClientParameters(ctx, cp, nil)
-		return conn, args, err
+		return netConn, args, err
 	}
 }
 
@@ -1606,21 +1625,23 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 	for _, tc := range testCases {
 		t.Run(tc.desc, func(t *testing.T) {
 
+			var netConn net.Conn
 			wg := sync.WaitGroup{}
 			wg.Add(1)
 			go func(query string) {
 				defer wg.Done()
-				ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+				ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
 				defer cancel()
 				url := fmt.Sprintf("%s&%s", baseURL, query)
 				c, connErr := pgx.Connect(ctx, url)
 				if connErr != nil {
+					t.Error(connErr)
 					return
 				}
 				// ignore the error because there is no answer from the server, we are
-				// interested in parsing session arguments only
-				_ = c.Ping(ctx)
-				// closing connection immediately, since getSessionArgs is blocking
+				// interested in parsing session arguments only. We close the connection
+				// immediately, since getSessionArgs is blocking.
+				_ = netConn.Close()
 				_ = c.Close(ctx)
 
 				select {
@@ -1632,7 +1653,8 @@ func TestParseClientProvidedSessionParameters(t *testing.T) {
 				}
 			}(tc.query)
 			// Wait for the client to connect and perform the handshake.
-			_, args, err := getSessionArgs(ln, true /* trustRemoteAddr */)
+			var args sql.SessionArgs
+			netConn, args, err = getSessionArgs(ln, true /* trustRemoteAddr */)
 			wg.Wait()
 			tc.assert(t, args, err)
 		})

